### PR TITLE
fix: shrink lock scope while populating cache

### LINKED ISSUE #7881: BUG:  performance degeneration if table cache enabled
**Summary**

Performance degeneration if table cache is enabled.

pls refer to the following comment for details

https://github.com/datafuselabs/databend/issues/7861#issuecomment-1257202309

### PR DESCRIPTION
**Pull request by dantengsky**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

cached meta reader should not hold the cache lock while reading data from storage 

**test scenario:**

table with 5050 segments

```
mysql> select segment_count from fuse_snapshot('db', 't') limit 1 \G
*************************** 1. row ***************************
segment_count: 5050
```

**statement being stressed:**

`explain select sum(id13) from db.t;`

with this patch:

cold start, hyperfine,  warmup 3, no table cache
~~~
  Time (mean ± σ):     284.9 ms ±  60.8 ms    [User: 3.3 ms, System: 1.3 ms]
  Range (min … max):   242.1 ms … 453.0 ms    11 runs
~~~

cold start, hyperfine,  warmup 3 , with table cache enabled
~~~
  Time (mean ± σ):      62.3 ms ±  10.4 ms    [User: 3.2 ms, System: 1.3 ms]
  Range (min … max):    44.9 ms …  87.9 ms    49 runs
~~~

Fixes #7881

### CODE REVIEW

File: src/query/storages/util/src/cached_reader.rs [hunk 1]
```diff
@@ -78,19 +78,23 @@ where L: Loader<T> + HasTenantLabel
                     cache_hit: false,
                     read_bytes: 0,
                 };
-                let cache = &mut cache.write().await;
-                match cache.get(location.as_ref()) {
-                    Some(item) => {
+
+                {
+                    let cache = &mut cache.write().await;
```
Comment (zhang2014) [inline src/query/storages/util/src/cached_reader.rs:83]: Let's use synchronous locks, which will let the compiler check lock holdings across async. Perhaps #7857 already contains these changes.

Comment (dantengsky) [reply src/query/storages/util/src/cached_reader.rs:83]: thanks bro! I am verifying the performance improvement of PR #7857 in EC2 + S3 setup. I'd like to drop this PR if test goes well ;)

Comment (dantengsky) [reply src/query/storages/util/src/cached_reader.rs:83]: @zhang2014  performance of PR #7857 is great for the scenario given is this PR's summary:

baseline : cold start, hyperfine, warmup 3, no table cache
~~~
Benchmark 1: mysql -uroot -P3307 -h 127.0.0.1 < explain.sql
  Time (mean ± σ):     268.6 ms ±  28.1 ms    [User: 3.0 ms, System: 1.6 ms]
  Range (min … max):   235.1 ms … 344.8 ms    30 runs
~~~  

cold start, hyperfine, warmup 3 , with table cache enabled

~~~
Benchmark 1: mysql -uroot -P3307 -h 127.0.0.1 < explain.sql
  Time (mean ± σ):      26.0 ms ±   0.5 ms    [User: 2.9 ms, System: 1.5 ms]
  Range (min … max):    25.2 ms …  28.0 ms    30 runs
~~~

**Comment by dantengsky** [comment]
Closed since PR https://github.com/datafuselabs/databend/pull/7857 also resolves the issue but in a more systematic/better way.

**Pull request closed without merging**